mobile_verifier/src/heartbeats.rs

```diff
@@ -145,27 +146,45 @@ pub struct SaveHeartbeatError(#[from] sqlx::Error);
 
 impl Heartbeat {
     pub async fn validate_heartbeats<'a>(
+        config_client: &'a Client,
         heartbeats: impl Stream<Item = CellHeartbeat> + 'a,
         epoch: &'a Range<DateTime<Utc>>,
-    ) -> impl Stream<Item = Self> + 'a {
+    ) -> impl Stream<Item = Result<Self, ClientError>> + 'a {
         let valid_epoch = epoch.start..(epoch.end + *MOBILE_INGEST_ROLL_TIME);
-        heartbeats.map(move |heartbeat_report| {
-            let (reward_weight, validity) =
-                match validate_heartbeat(&heartbeat_report, &valid_epoch) {
-                    Ok(cell_type) => {
-                        let reward_weight = cell_type.reward_weight();
-                        (reward_weight, proto::HeartbeatValidity::Valid)
-                    }
-                    Err(validity) => (dec!(0), validity),
-                };
-            Heartbeat {
-                hotspot_key: heartbeat_report.pubkey.clone(),
-                reward_weight,
-                cbsd_id: heartbeat_report.cbsd_id.clone(),
-                timestamp: heartbeat_report.timestamp.naive_utc(),
-                validity,
-            }
-        })
+        heartbeats
+            .map(move |heartbeat_report| {
+                let (reward_weight, validity) =
+                    match validate_heartbeat(&heartbeat_report, &valid_epoch) {
```

Comment:
Would it make sense to put the gateway owner check in the `validate_heartbeat` function itself. It already checks other errors (not_operational, bad_cbsd_id and outside_range). You could probably pass in the resolver to `validate_heartbeat` function and have it do the ownership check. That way the subsequent `.then` won't be necessary.

Reply:
I considered that, but I liked how that function returned a Result. But I'll look back at in and see if there's a good way of combining them